File: src/reducers/map.js
import T from "../constants/ActionTypes";
import mapConst from "../constants/Map";
import parseUrl from "../util/parseUrl";

const initialState = {
  zoom: mapConst.DEFAULT_ZOOM,
  center: mapConst.DEFAULT_CENTER,
  marker: null,
  bbox: mapConst.DEFAULT_BBOX,
  ownPosition: null,
  ownPositionCancelled: false,
  waiting_for_center_of: null
};

export default (state = initialState, action = {}) => {

  var newState, point;
  const { payload } = action;

  switch (action.type) {
    case T.UPDATE_STATE_FROM_URL:
      const { center, zoom } = parseUrl(payload).params;
      const mapCenter = center ? {
        lat: parseFloat(center.split(',')[0]),
        lng: parseFloat(center.split(',')[1])
      } : null;

      return {
        ...state,
        center: mapCenter || state.center || mapConst.DEFAULT_CENTER,
        zoom: parseFloat(zoom) || state.zoom || mapConst.DEFAULT_ZOOM
    }
    case T.CLOSE_NEW_ENTRY:
    case T.SHOW_NEW_ENTRY:
      return {
        ...state,
        marker: null
      };
    case T.SET_MARKER:
      if (action.manual) {
        return {
          ...state,
          marker: action.payload
        };
      } else {
        return {
          ...state,
          marker: action.payload,
          center: action.payload
        };
      }
      break;
    case T.EDIT_CURRENT_ENTRY:
      point = {
        lat: action.payload.lat,
        lng: action.payload.lng
      };
      return {
        ...state,
        marker: point,
        center: point
      };
    case T.SET_MAP_CENTER:
      if ((typeof action.payload) === "string") {
        // payload is an entry ID
        return {
          ...state,
          waiting_for_center_of: action.payload
        };

      } else {
        return {
          ...state,
          center: {
            lat: parseFloat(payload.lat),
            lng: parseFloat(payload.lng)
          }
        };
      }

    case T.ENTRIES_RESULT:
    case T.SEARCH_RESULT_EVENTS:
      if ((payload != null) && (payload.length > 0) && (state.waiting_for_center_of != null)) {
        var o = {};
        if (Array.isArray(payload)) {
          payload.filter(e => e != null)
           .forEach(e => { o[e.id] = e; });
        } else {
          o[payload.id] = payload;
        }
        if(o[state.waiting_for_center_of]){
          return{
            ...state,
            center: {
              lat: o[state.waiting_for_center_of].lat,
              lng: o[state.waiting_for_center_of].lng
            },
            waiting_for_center_of: null
          }
        } else {
          return state
        }
      } else{
        return state;
      }
    case T.SET_ZOOM:
      return {
        ...state,
        zoom: action.payload
      };
    case T.SET_BBOX:
      return {
        ...state,
        bbox: action.payload
      };

    case T.ENTRY_RESULT:
      if (action.error) {
        return state;
      }
      const e_id = state.waiting_for_center_of;
      const e = action.payload[e_id];
      if (e) {
        const { lat, lng } = e;
        return {
          ...state,
          waiting_for_center_of: null,
          center: { lat, lng }
        };
      }
      return state;

    case T.NEW_ENTRY_RESULT:
      if (!action.error) {
        return {
          ...state,
          marker: null
        };
      }
      return state;
    case T.SHOW_OWN_POSITION:
      return {
        ...state,
        ownPositionCancelled: false
      };
    case T.CANCEL_OWN_POSITION:
      return {
        ...state,
        ownPositionCancelled: true
      };
    case T.OWN_POSITION_RESULT:
      if (action.payload) {
        const newState = {
          ...state,
          ownPosition: {
            lat: action.payload.coords.latitude,
            lng: action.payload.coords.longitude
          }
        };
        if (!state.ownPositionCancelled) {
          return {
            ...newState,
            center: newState.ownPosition
          };
        } else {
          return newState;
        }
      } else {
        return state;
      }
      break;

    default:
      return state;
  }
};

File: src/reducers/map.rs
// TODO: import T from "../constants/ActionTypes";
// TODO: import mapConst from "../constants/Map";
// TODO: import parseUrl from "../util/parseUrl";
// TODO: 
// TODO: const initialState = {
// TODO:   zoom: mapConst.DEFAULT_ZOOM,
// TODO:   center: mapConst.DEFAULT_CENTER,
// TODO:   marker: null,
// TODO:   bbox: mapConst.DEFAULT_BBOX,
// TODO:   ownPosition: null,
// TODO:   ownPositionCancelled: false,
// TODO:   waiting_for_center_of: null
// TODO: };
// TODO: 
// TODO: module.exports = (state = initialState, action = {}) => {
// TODO: 
// TODO:   var newState, point;
// TODO:   const { payload } = action;
// TODO: 
// TODO:   switch (action.type) {
// TODO:     case T.UPDATE_STATE_FROM_URL:
// TODO:       const { center, zoom } = parseUrl(payload).params;
// TODO:       const mapCenter = center ? {
// TODO:         lat: parseFloat(center.split(',')[0]),
// TODO:         lng: parseFloat(center.split(',')[1])
// TODO:       } : null;
// TODO: 
// TODO:       return {
// TODO:         ...state,
// TODO:         center: mapCenter || state.center || mapConst.DEFAULT_CENTER,
// TODO:         zoom: parseFloat(zoom) || state.zoom || mapConst.DEFAULT_ZOOM
// TODO:     }
// TODO:     case T.CLOSE_NEW_ENTRY:
// TODO:     case T.SHOW_NEW_ENTRY:
// TODO:       return {
// TODO:         ...state,
// TODO:         marker: null
// TODO:       };
// TODO:     case T.SET_MARKER:
// TODO:       if (action.manual) {
// TODO:         return {
// TODO:           ...state,
// TODO:           marker: action.payload
// TODO:         };
// TODO:       } else {
// TODO:         return {
// TODO:           ...state,
// TODO:           marker: action.payload,
// TODO:           center: action.payload
// TODO:         };
// TODO:       }
// TODO:       break;
// TODO:     case T.EDIT_CURRENT_ENTRY:
// TODO:       point = {
// TODO:         lat: action.payload.lat,
// TODO:         lng: action.payload.lng
// TODO:       };
// TODO:       return {
// TODO:         ...state,
// TODO:         marker: point,
// TODO:         center: point
// TODO:       };
// TODO:     case T.SET_MAP_CENTER:
// TODO:       if ((typeof action.payload) === "string") {
// TODO:         // payload is an entry ID
// TODO:         return {
// TODO:           ...state,
// TODO:           waiting_for_center_of: action.payload
// TODO:         };
// TODO: 
// TODO:       } else {
// TODO:         return {
// TODO:           ...state,
// TODO:           center: {
// TODO:             lat: parseFloat(payload.lat),
// TODO:             lng: parseFloat(payload.lng)
// TODO:           }
// TODO:         };
// TODO:       }
// TODO: 
// TODO:     case T.ENTRIES_RESULT:
// TODO:     case T.SEARCH_RESULT_EVENTS:
// TODO:       if ((payload != null) && (payload.length > 0) && (state.waiting_for_center_of != null)) {
// TODO:         var o = {};
// TODO:         if (Array.isArray(payload)) {
// TODO:           payload.filter(e => e != null)
// TODO:            .forEach(e => { o[e.id] = e; });
// TODO:         } else {
// TODO:           o[payload.id] = payload;
// TODO:         }
// TODO:         if(o[state.waiting_for_center_of]){
// TODO:           return{
// TODO:             ...state,
// TODO:             center: {
// TODO:               lat: o[state.waiting_for_center_of].lat,
// TODO:               lng: o[state.waiting_for_center_of].lng
// TODO:             },
// TODO:             waiting_for_center_of: null
// TODO:           }
// TODO:         } else {
// TODO:           return state
// TODO:         }
// TODO:       } else{
// TODO:         return state;
// TODO:       }
// TODO:     case T.SET_ZOOM:
// TODO:       return {
// TODO:         ...state,
// TODO:         zoom: action.payload
// TODO:       };
// TODO:     case T.SET_BBOX:
// TODO:       return {
// TODO:         ...state,
// TODO:         bbox: action.payload
// TODO:       };
// TODO: 
// TODO:     case T.ENTRY_RESULT:
// TODO:       if (action.error) {
// TODO:         return state;
// TODO:       }
// TODO:       const e_id = state.waiting_for_center_of;
// TODO:       const e = action.payload[e_id];
// TODO:       if (e) {
// TODO:         const { lat, lng } = e;
// TODO:         return {
// TODO:           ...state,
// TODO:           waiting_for_center_of: null,
// TODO:           center: { lat, lng }
// TODO:         };
// TODO:       }
// TODO:       return state;
// TODO: 
// TODO:     case T.NEW_ENTRY_RESULT:
// TODO:       if (!action.error) {
// TODO:         return {
// TODO:           ...state,
// TODO:           marker: null
// TODO:         };
// TODO:       }
// TODO:       return state;
// TODO:     case T.SHOW_OWN_POSITION:
// TODO:       return {
// TODO:         ...state,
// TODO:         ownPositionCancelled: false
// TODO:       };
// TODO:     case T.CANCEL_OWN_POSITION:
// TODO:       return {
// TODO:         ...state,
// TODO:         ownPositionCancelled: true
// TODO:       };
// TODO:     case T.OWN_POSITION_RESULT:
// TODO:       if (action.payload) {
// TODO:         const newState = {
// TODO:           ...state,
// TODO:           ownPosition: {
// TODO:             lat: action.payload.coords.latitude,
// TODO:             lng: action.payload.coords.longitude
// TODO:           }
// TODO:         };
// TODO:         if (!state.ownPositionCancelled) {
// TODO:           return {
// TODO:             ...newState,
// TODO:             center: newState.ownPosition
// TODO:           };
// TODO:         } else {
// TODO:           return newState;
// TODO:         }
// TODO:       } else {
// TODO:         return state;
// TODO:       }
// TODO:       break;
// TODO: 
// TODO:     default:
// TODO:       return state;
// TODO:   }
// TODO: };
